### PR TITLE
Minimize Make-Style Dependencies

### PR DESCRIPTION
**Pull request by CodaFi**
When -emit-dependencies is switched on, the Frontends are each tasked
with emitting Make-style dependencies files describing the full Swift
compilation graph and any external dependencies picked up by the Clang
Importer. Each output path was given a make target, and each target was
given a full copy of the dependencies of the swift compilation session.

This is a tremendous amount of duplication of a tremendous amount of
dependencies.  There can often be hundreds of dependencies for a single
target, and each output file's target would recieve its own copy. This
puts a significant strain on the build system, which is parsing a linear
number of these dependency files in incremental mode. (Given the build
system operates at the level of driver invocations, it's questionable
whether we even need to emit these files with content at all, but that's
a separate issue). When each output file adds all of its dependencies,
that means we have a quadratic increase in the parsing load.

Aim to compress these make-style dependency files by nominating
a primary output as a most-preferred target that recieves the entire
compilation graph.  The remaining targets, rather than copy the build
dependencies, simply depend on that target being rebuilt.

This allows us to cut down the size of make-style dependencies by
(1/# of outputs).  When multiplied by the number of files in a large
incremental codebase, as in SR-11927, the savings can be as much as
50% of incremental build time.

We can reduce this further by nominating a single frontend job to be the
Blessed One That Emits The Dependencies since Make requires we emit
a complete graph (as in K^n) so everybody sees the same internal
dependencies, and there's no way for a user to customize
individual frontend invocations as with Clang so everybody should see
the same external dependencies.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
Uh oh, this is gonna fail a Driver test.  I'll get to it tomorrow.

File: lib/FrontendTool/FrontendTool.cpp [hunk 1]
```diff
@@ -135,6 +135,93 @@ swift::frontend::utils::escapeForMake(StringRef raw,
...
+        // source files are going to map to the same main output file.
+        if (seenTargets.insert(input.outputFilenameRef()).second) {
+            preferredOutputs.push_back(&input.outputFilenameRef());
+        }
+        // Don't early-exit.
+        return false;
+      });
+    } else {
+      preferredOutputs.push_back(&input.outputFilenameRef());
+    }
```
Comment (davidungar) [inline lib/FrontendTool/FrontendTool.cpp]: I find this easier to read when the short branch, here the else case, comes first. That way I don't have to match up the brace before the "else" with the appropriate "if".

File: lib/FrontendTool/FrontendTool.cpp [hunk 2]
```diff
@@ -135,6 +135,93 @@ swift::frontend::utils::escapeForMake(StringRef raw,
...
+  mainTarget(*firstNonEmptyTarget);
+
+  // Now scan forward to apply the supplementary function
+  // to the remaining targets.
+  for (++it; it != std::end(preferredOutputs); ++it) {
+    if (!(*it)->empty()) {
+      remainingTarget(**it, *firstNonEmptyTarget);
+    }
+  }
+}
```
Comment (davidungar) [inline lib/FrontendTool/FrontendTool.cpp]: This function takes up almost the whole of my screen height (40") without the comment. Could it be broken up into about three pieces? Your excellent comments illustrate that each piece does have a cogent separable thought to it.

File: lib/FrontendTool/FrontendTool.cpp [hunk 3]
```diff
@@ -135,6 +135,93 @@ swift::frontend::utils::escapeForMake(StringRef raw,
...
+  auto it = llvm::find_if(preferredOutputs, [](const std::string *str) {
+    return !str->empty();
+  });
+
+  // No targets?  There's nothing to do here.
+  if (it == std::end(preferredOutputs))
+    return;
+
+  // Apply the function to the most preferred target.
+  const std::string *const firstNonEmptyTarget(*it);
```
Comment (davidungar) [inline lib/FrontendTool/FrontendTool.cpp]: Nice "consting", BTW.

File: test/Driver/minimal_make_dependencies.swift [hunk 4]
```diff
@@ -0,0 +1,63 @@
+// Try WMO - we need to emit targets for each primary object file.
```
Comment (davidungar) [inline test/Driver/minimal_make_dependencies.swift]: Would be nice to add a comment at the top of the file summarizing what is being tested.

File: lib/FrontendTool/FrontendTool.cpp [hunk 5]
```diff
@@ -135,6 +135,93 @@ swift::frontend::utils::escapeForMake(StringRef raw,
...
+// The final structure of the dependency file is therefore:
+//
+// /Path/To/Most/Preferred/Output/File.o:
+//   <Big Ol' List Of Files...>
+// /Path/To/Supplementary/Output/File.swiftdoc:
+//   /Path/To/Most/Preferred/Output/File.o
+// /Path/To/Supplementary/Output/File.swiftsourceinfo:
+//   /Path/To/Most/Preferred/Output/File.o
+// ...
+static void forEachMinimalMakeStyleTarget(
```
Comment (davidungar) [inline lib/FrontendTool/FrontendTool.cpp]: I like the structure of using a functional interface here.

**Review by davidungar — CHANGES_REQUESTED** [review]
Very nice! Great idea, done very quickly. The one comment I feel strongly about is that the one large function be broken up. Take or leave the rest. Thanks in advance!

**Comment by DougGregor** [comment]
@swift-ci please test Windows

**Comment by CodaFi** [comment]
@swift-ci please test windows

**Comment by CodaFi** [comment]
@swift-ci please test windows

**Comment by CodaFi** [comment]
@swift-ci please test windows

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test windows

**Comment by CodaFi** [comment]
Hmph, need to update the check line to make Windows happy...

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test Windows

**Comment by CodaFi** [comment]
@swift-ci please test Windows

**Comment by CodaFi** [comment]
@swift-ci please test Windows

**Comment by CodaFi** [comment]
@swift-ci please test Windows

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
If this were still open I would approve it. Looks better!

**Pull request closed without merging**